Automate end-to-end permission testing

kaankabalak opened the pull request:
This PR sets up and uses the [TestCafe framework](https://testcafe.io/) to automate end-to-end permission testing.

To summarize, these tests involve the creation of users and policies via `mc` and uses them to try out various permission related operations. Therefore it requires multiple commands to initialize and clean them up.

The entirety of the process can be run using the `env bash /portal-ui/tests/scripts/permissions.sh` command; provided that:
- TestCafe is installed globally (using the `npm i -g testcafe` command)
- The user is on the `console` directory
- Before the tests are run, the front-end needs to be built (using the `yarn build` command)
- MinIO Server (on port 9000), MinIO Console (on port 9090) are both running

Fixes miniohq/engineering#254

kaankabalak commented on .github/workflows/permissions.yml:
```diff
@@ -0,0 +1,42 @@
+on:
+  pull_request:
+    branches:
+      - master
+  push:
+    branches:
+      - master
+
+jobs:
+  permissions:
+    name: Front-End Permission Tests
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        go-version: [1.17.x]
+        os: [ubuntu-latest]
+    steps:
+      - name: Set up Go ${{ matrix.go-version }} on ${{ matrix.os }}
+        uses: actions/setup-go@v2
+        with:
+          go-version: ${{ matrix.go-version }}
+        id: go
+
+      - name: Check out code into the Go module directory
+        uses: actions/checkout@v2
+
+      - name: Build Console on ${{ matrix.os }}
+        env:
+          GO111MODULE: on
+          GOOS: linux
+        run: |
+          make console
+
+      - name: Install modules
+        working-directory: ./portal-ui
+        run: yarn
+
+      - name: Run permission tests
+        run: |
+          make test-permissions
```
While writing the other workflow .yml files, how did you make sure that the actions ran without any issues? I was trying to use the [nektos/act](https://github.com/nektos/act) library, but I faced some problems with the make commands not being recognized. I would appreciate if you could point me towards the right direction. Thanks in advance!

@dvaldivia @Alevsk @bexsoft

kaankabalak replied:
GitHub Actions related issues have been fixed 👍

kaankabalak commented on Makefile:
```diff
@@ -68,6 +68,11 @@ test-integration:
 	@(docker stop minio)
 
+test-permissions:
+	@(docker run -d --name minio --rm -p 9000:9000 quay.io/minio/minio:latest server /data{1...4})
```
As the MinIO Server is running on port 9000, regardless of it being in a Docker container or not, we are sure that it will be recognized by the tests pointing to that port, right?

dvaldivia replied:
It should, 9000 on localhost is the default minio server address console tries to connect

kaankabalak commented on portal-ui/tests/constants/timestamp.txt:
```diff
@@ -0,0 +1 @@
+1641954496
```
This is a placeholder timestamp, it will be replaced with the Unix timestamp at the time when the user runs the /portal-ui/tests/scripts/permissions.sh script.

prakashsvmx commented:
@kaankabalak  please update all the new files to have year as 2022 in the license header.

kaankabalak commented on portal-ui/tests/scripts/permissions.sh:
```diff
@@ -0,0 +1,127 @@
+}
+
+create_policies() {
+  mc admin policy add minio bucketassignpolicy-$TIMESTAMP $(PWD)/portal-ui/tests/policies/bucketAssignPolicy.json
```
From the test that has just run, it seems like there's a problem with the recognition of `$(PWD)`, will take a look at it. Please let me know if know what might be causing this. Thanks!

kaankabalak replied:
Fixed

kaankabalak commented:
> @kaankabalak please update all the new files to have year as 2022 in the license header.

Will do @prakashsvmx

kaankabalak commented:
> @kaankabalak please update all the new files to have year as 2022 in the license header.

Done

kaankabalak commented:
I have just realized that we had some UI changes after I have rebased the branch, I will update the tests shortly. You can hold off reviewing the PR until then.

@dvaldivia @prakashsvmx @bexsoft @Alevsk @jinapurapu @adfost @cniackz

kaankabalak commented:
> I have just realized that we had some UI changes after I have rebased the branch, I will update the tests shortly. You can hold off reviewing the PR until then.

@dvaldivia @prakashsvmx @bexsoft @Alevsk @jinapurapu @adfost @cniackz

I have fixed the tests that were failing after recent UI changes, therefore you can now review the sections that involve the tests. However, I will need some help making sure the action runs smoothly, and am currently taking a look at the `PWD` issue that I previously mentioned when the action is being run.

kaankabalak commented:
Even though some tests within the Permission tests didn't pass, Github says that the checks were successful. So, I will be taking a look to see why this might be happening.

bexsoft commented on portal-ui/src/screens/Console/Buckets/ListBuckets/Objects/ListObjects/ListObjects.tsx:
```diff
@@ -1219,14 +1219,6 @@ const ListObjects = ({
-                    style={{ display: "none" }}
-                    ref={folderUpload}
-                  />
                 </SecureComponent>
```
Please don't remove these lines, this is required for folder upload functionality.

kaankabalak replied:
Could you explain how having two inputs work here? I don't think I understand it fully and when I tested out the upload folder functionality with a single input it worked without any issues both when uploading from the file navigator as well as dragging and dropping a file. Seems like we just have a single input for the file upload functionality, so I just wanted to ask. Thanks in advance! 👍

dvaldivia replied:
I think this change is fine, it's only putting them inside a fragment

bexsoft replied:
If you click on upload folder button it may not allow you to select a folder as this second input tag adds extra properties in the folderUpload ref, these properties cannot be added in the normal select file as file selector only allows to use one mode (select files or select folders)

kaankabalak replied:
Thanks @bexsoft for the explanation, I have kept the second input element and enclosed the child component for the SecureComponent inside a Fragment component.

kaankabalak commented:
> Even though some Permission tests didn't pass, Github said that the test suite ran successfully. I will be taking a look to see why this might be happening.

Seems like this is fixed by using the TestCafe GitHub action, however we now need to make sure the front-end app is running before the test suite starts.

kaankabalak commented:
I fixed the issue where the front-end app didn't load in time for the before the permission tests started, I will now update the tests based on the latest UI changes.

kaankabalak commented:
The GitHub Actions related issues have all been fixed, I would appreciate if you could review the PR. Thanks once again! 👍 

@dvaldivia @prakashsvmx @bexsoft @Alevsk @jinapurapu @adfost @cniackz

dvaldivia reviewed (dismissed):
LGTM.

dvaldivia commented:
@kaankabalak I've been checking this since the first version, I think it's all clear, my only feedback would be to split the `elements.ts` file into smaller chunks, perhaps even per screen, to make it easier for devs to add more elements and tests in the future, but this is an amazing start

harshavardhana commented:
> @kaankabalak I've been checking this since the first version, I think it's all clear, my only feedback would be to split the `elements.ts` file into smaller chunks (maybe in another PR), perhaps even per screen, to make it easier for devs to add more elements and tests in the future, but this is an amazing start

Agreed finally some nice :cake:

dvaldivia commented:
@kaankabalak there's a merge conflict

kaankabalak commented:
> @kaankabalak there's a merge conflict

Fixed the conflict @dvaldivia

kaankabalak commented:
Seems like a test is randomly failing only on GitHub Actions, will investigate this and send a fix for this in a future PR.

bexsoft reviewed (approved):
LGTM
